src/main/kotlin/org/jitsi/nlj/stats/DelayStats.kt

```diff
@@ -16,38 +16,132 @@
 
-import java.util.Arrays
 import java.util.concurrent.atomic.AtomicLong
 import java.util.concurrent.atomic.LongAdder
 import org.jitsi.nlj.PacketInfo
+import org.jitsi.nlj.util.OrderedJsonObject
+import org.jitsi.utils.maxAssign
+import java.lang.IllegalArgumentException
 
-open class DelayStats {
+open class DelayStats(thresholdsNoMax: LongArray = defaultThresholds) {
+    init {
+        if (!thresholdsNoMax.contentEquals(thresholdsNoMax.sortedArray())) {
+            throw IllegalArgumentException("Thresholds must be sorted: ${thresholdsNoMax.joinToString()}")
+        }
+    }
     private val totalDelayMs = LongAdder()
     private val totalCount = LongAdder()
-    val averageDelay: Double
+    private val averageDelayMs: Double
         get() = totalDelayMs.sum() / totalCount.sum().toDouble()
-    val maxDelayMs = AtomicLong(0)
-
-    /* TODO: make thresholds configurable */
-    val thresholds = longArrayOf(2, 5, 20, 50, 200, 500, 1000)
-    val thresholdCounts = Array(thresholds.size + 1) { LongAdder() }
+    private val maxDelayMs = AtomicLong(0)
+    private val buckets = Buckets(thresholdsNoMax)
 
     fun addDelay(delayMs: Long) {
         if (delayMs >= 0) {
             totalDelayMs.add(delayMs)
-            if (delayMs > maxDelayMs.get()) {
-                maxDelayMs.set(delayMs)
-            }
+            maxDelayMs.maxAssign(delayMs)
             totalCount.increment()
+            buckets.addDelay(delayMs)
+        }
+    }
+
+    fun toJson() = OrderedJsonObject().apply {
+        val snapshot = snapshot
+        put("average_delay_ms", snapshot.averageDelayMs)
+        put("max_delay_ms", snapshot.maxDelayMs)
+        put("total_count", snapshot.totalCount)
+
+        put("buckets", snapshot.buckets.toJson())
+    }
+
+    val snapshot: Snapshot
+        get() = Snapshot(
+            averageDelayMs = averageDelayMs,
+            maxDelayMs = maxDelayMs.get(),
+            totalCount = totalCount.sum(),
+            buckets = buckets.snapshot
+        )
+
+    data class Snapshot(
+        val averageDelayMs: Double,
+        val maxDelayMs: Long,
+        val totalCount: Long,
+        val buckets: Buckets.Snapshot
+    )
+
+    companion object {
+        val defaultThresholds = longArrayOf(2, 5, 20, 50, 200, 500, 1000)
+    }
+}
+
+class Buckets(thresholdsNoMax: LongArray) {
+    private val thresholds = longArrayOf(*thresholdsNoMax, Long.MAX_VALUE)
+    private val thresholdCounts = Array(thresholds.size + 1) { LongAdder() }
+    val snapshot: Snapshot
+        get() {
+            val bucketCounts = Array(thresholds.size) { i -> Pair(thresholds[i], thresholdCounts[i].sum()) }
+
+            var p99 = Long.MAX_VALUE
+            var p999 = Long.MAX_VALUE
+            var sum: Long = 0
+            val totalCount = bucketCounts.map { it.second }.sum()
+            bucketCounts.forEach {
+                sum += it.second
+                if (it.first < p99 && sum > 0.99 * totalCount) p99 = it.first
```

Comment:
should we do the checks for there being enough data for at least p99 before doing this loop?

Reply:
The checks for p99 and p999 are different, so the code would get hairy. This only executes when the stats are queried and the number of thresholds is small, so it's not a perf issue.

Reply:
Right, but if there are less than 100 then we know we don't need to do the loop at all.  I don't think it's a big deal though.